nexus/src/app/project.rs

```diff
@@ -85,8 +85,10 @@ impl super::Nexus {
                 saga_params,
             )
             .await?;
-        let db_project = saga_outputs
-            .lookup_node_output::<db::model::Project>("project")
+        let (_authz_project, db_project) = saga_outputs
+            .lookup_node_output::<(authz::Project, db::model::Project)>(
```

Comment:
Question: is it _safe_ to store `authz` objects in the Saga like this? I (think I) understand the motivation to do so, I'm just wondering (probably too informally) if they could somehow be extracted and misused, or if having `authz` instances be deserializable opens the door to attacks like starting with a fake serialization and getting a real but unauthorized `authz` object from that.

Reply:
I discussed something similar with @davepacheco last week: https://github.com/oxidecomputer/omicron/pull/2060#discussion_r1054890025

Let's suppose we don't do this, and that we refuse to serialize any authz objects. In that case, every saga step acting on an object would need to look it up again, and re-authorize. One particularly weird fall-out of this is that changing authorizations for a particular user mid-saga could result in some gnarly edge-cases. For example:

- Suppose a user is authorized to create / modify disks
- Suppose the user is partway through a disk-creation saga
- Suppose that saga *fails*, for whatever reason (Out of space? Crucible bug? Whatever)
- Suppose that **concurrently, the authorizations granted to that user are modified to disallow disk modification**.
- Now, when unwinding the saga operations, we **cannot actually remove the partially-created disk, because our authorization check would fail**. 

This seems quite bad to me. I understand that there are other latent concerns around authz objects, such as "can we guarantee they are constructed safely", but an inability to pass them across saga nodes implies that "the whole authz world might change" between saga operations, which seems like it could make unwinding fail arbitrarily.

Reply:
Thanks; I always feel better knowing I'm thinking along similar lines as Dave :smile: Your arguments make total sense from a practical standpoint, and the concerns raised are so far entirely hypothetical. But I will paste a quote from the linked comment here just for emphasis:
> as long as we only ever serialize these into saga state (and not elsewhere in the database, for example, or over HTTP interfaces), I think you're right that this is fine.

One of the things that I take from your example above is that using sagas implies a kind of temporal snapshotting, rather like a closure: when we create a saga, we need to capture the state associated with the relevant database objects at that instant in time. That necessarily includes authorization for the included queries/modifications, and I think you're right that using a new context to undo old actions seems semantically wrong. So, all this to say: thanks for considering the question, I've come around and see now this is the only way to properly handle this with the current machinery.